File: 先序遍历中序遍历和后续遍历.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node { // 树节点数据结构
    int left = 0;
    int right = 0;
};

int main() {
    int n, start;
    cin >> n;
    vector<Node> node(n + 1);
    vector<int> in(n + 1);
    for (int i = 1; i < n; ++i) { // 建树
        int a, b;
        cin >> a >> b;
        in[b]++;
        if (node[a].left == 0 and node[a].right == 0) { // 若当前节点 a 没孩子，则当作“父节点仅有一个孩子”
            if (a < b) {
                node[a].left = b;
            } else {
                node[a].right = b;
            }
        } else if (node[a].left) { // 若当前节点 a 有一个左孩子，做出相对应的调整
            if (node[a].left > b) {
                node[a].right = node[a].left;
                node[a].left = b;
            } else {
                node[a].right = b;
            }
        } else { // 若当前节点 a 有一个右孩子，做出相对应的调整
            if (node[a].right < b) {
                node[a].left = node[a].right;
                node[a].right = b;
            } else {
                node[a].left = b;
            }
        }
    }
    for (int i = 1; i <= n; ++i) {
        if (in[i] == 0) {
            start = i;
            break;
        }
    }

    function<void(int)> pre_search = [&](int now) {
        cout << now << " ";
        if (node[now].left) {
            pre_search(node[now].left);
        }
        if (node[now].right) {
            pre_search(node[now].right);
        }
    };

    function<void(int)> mid_search = [&](int now) {
        if (node[now].left) {
            mid_search(node[now].left);
        }
        cout << now << " ";
        if (node[now].right) {
            mid_search(node[now].right);
        }
    };

    function<void(int)> rear_search = [&](int now) {
        if (node[now].left) {
            rear_search(node[now].left);
        }
        if (node[now].right) {
            rear_search(node[now].right);
        }
        cout << now << " ";
    };

    pre_search(start);
    cout << endl;
    mid_search(start);
    cout << endl;
    rear_search(start);
}

File: 先序遍历中序遍历和后续遍历.py
class Node:
    def __init__(self):
        self.left = 0
        self.right = 0

n = int(input())
node = [Node() for _ in range(n + 1)]
IN = [0] * (n + 1) # 入度数组，用于寻找根节点

# 读取 n-1 条边并建树
for _ in range(n - 1):
    a, b = map(int, input().split())
    IN[b] += 1
    if node[a].left == 0 and node[a].right == 0:
        if a < b:
            node[a].left = b
        else:
            node[a].right = b
    elif node[a].left != 0:
        if node[a].left > b:
            # 原左孩子移到右边，新节点 b 成为左孩子
            node[a].right = node[a].left
            node[a].left = b
        else:
            node[a].right = b
    else: 
        if node[a].right < b:
            # 原右孩子移到左边，新节点 b 成为右孩子
            node[a].left = node[a].right
            node[a].right = b
        else:
            node[a].left = b

start = 0
for i in range(1, n + 1):
    if IN[i] == 0:
        start = i
        break

def pre_search(now):
    print(now, end = " ")
    if node[now].left != 0:
        pre_search(node[now].left)
    if node[now].right != 0:
        pre_search(node[now].right)

# 中序遍历 (Left -> Root -> Right)
def mid_search(now):
    if node[now].left != 0:
        mid_search(node[now].left)
    print(now, end = " ")
    if node[now].right != 0:
        mid_search(node[now].right)

# 后序遍历 (Left -> Right -> Root)
def rear_search(now):
    if node[now].left != 0:
        rear_search(node[now].left)
    if node[now].right != 0:
        rear_search(node[now].right)
    print(now, end = " ")

pre_search(start)
print()
mid_search(start)
print()
rear_search(start)
print()
